### PR TITLE
Remove redundant return of the error 

### PR DESCRIPTION
**Pull request by ssinha20**
existing issues in crypto_test.go solved.

### CODE REVIEW

**Comment by harche** [comment]
@ssinha20 thanks for the PR. 

Can you squash the commits?

File: pkg/crypto/crypto_test.go [hunk 1]
```diff
@@ -111,3 +112,85 @@ func decryptConfigMap(configMap string, keyPath string, noncePath string) string
...
+
+        //Encode the Key
+        keyEnc := b64.StdEncoding.EncodeToString(key)
+        if err != nil {
+                t.Error(err)
+        }
+
+        //encode wrong key
+        wrongkeyEnc := b64.StdEncoding.EncodeToString(wrongkey)
+        if err != nil {
```
Comment (harche) [inline pkg/crypto/crypto_test.go]: Which error is this? `EncodeToString` doesn't return error. https://golang.org/pkg/encoding/base64/#example_Encoding_EncodeToString

File: pkg/crypto/crypto_test.go [hunk 2]
```diff
@@ -111,3 +112,85 @@ func decryptConfigMap(configMap string, keyPath string, noncePath string) string
...
+
+        wrongkey := make([]byte, 32)
+        if _, err := io.ReadFull(rand.Reader, wrongkey); err != nil {
+                t.Error(err)
+        }
+
+
+        //Encode the Key
+        keyEnc := b64.StdEncoding.EncodeToString(key)
+        if err != nil {
```
Comment (harche) [inline pkg/crypto/crypto_test.go]: Which error is this? EncodeToString doesn't return error. https://golang.org/pkg/encoding/base64/#example_Encoding_EncodeToString

File: pkg/crypto/crypto.go [hunk 3]
```diff
@@ -66,9 +66,9 @@ func EncryptConfigMap(configMap []byte, keyPath string, noncePath string) (encCo
 
 // GetConfigMapKeys returns the keys used for encrypting the configMap
 func GetConfigMapKeys(keyPath string, noncePath string) (symmKey []byte, nonce []byte, err error) {
-	if symmetricKey != nil && symmKeyNonce != nil {
+	/*if symmetricKey != nil && symmKeyNonce != nil {
```
Comment (harche) [inline pkg/crypto/crypto.go]: Why is this change required? This check exists to return an existing key and nonce.

File: pkg/crypto/crypto_test.go [hunk 4]
```diff
@@ -111,3 +112,85 @@ func decryptConfigMap(configMap string, keyPath string, noncePath string) string
...
+        encConfigMap, err := EncryptConfigMap([]byte(testString), keyPath, noncePath)
+
+        if err != nil {
+                t.Errorf("Failed to encrypt configMap %v", err)
+        }
+
+        wrongdecryptedConfigMap := decryptConfigMap(encConfigMap, wrongkeyPath, noncePath)
+        fmt.Printf("What is being returned by decrypt function on getting a wrong key = %s", wrongdecryptedConfigMap)
+
+        if testString != wrongdecryptedConfigMap {
```
Comment (harche) [inline pkg/crypto/crypto_test.go]: If you use the incorrect key to decrypt the configmap, won't this condition be always true? Also, `decryptConfigMap` should return `err` instead of panicking (https://github.com/IBM/raksh/blob/6b2e4df4e38bd747e098cd8c0f8119c037cde1bc/pkg/crypto/crypto_test.go#L90). That will make it easier to handle it here.

**Review by harche — COMMENTED** [review]
please use `gofmt` on the changes.

File: pkg/crypto/crypto_test.go (hunk 1 above)
Comment (ssinha20) [reply pkg/crypto/crypto_test.go]: noted. will correct this. this is followed in existing bucket. will correct them too.
//Encode the Key
	keyEnc := b64.StdEncoding.EncodeToString(key)
	if err != nil {
		t.Error(err)
	}

File: pkg/crypto/crypto.go (hunk 3 above)
Comment (ssinha20) [reply pkg/crypto/crypto.go]: Correct. But the issue is, these are global variables and passed values are not checked against them. So for example , in decrypt, if wrong key is passed (any non null value), in stead of returning the passed value, global value is returned. So I prefer to have check in stead of just returning. alternate way can be, compare global value and passed value.

File: pkg/crypto/crypto_test.go (hunk 4 above)
Comment (ssinha20) [reply pkg/crypto/crypto_test.go:191]: I agree. That's exactly what I had in mind, when I wanted to write negative scenarios. As it was something pre-existing, I did not change it. However, I will discuss this with my tech lead and update further on this.

Comment (ssinha20) [reply pkg/crypto/crypto_test.go]: Thanks for the suggestions. Do you prefer I address this in this PR or create a separate PR for this

File: pkg/crypto/crypto_test.go (hunk 1 above)
Comment (ssinha20) [reply pkg/crypto/crypto_test.go]: Also, Do you prefer I address this in this PR or create a separate PR for this? Please suggest.

File: pkg/crypto/crypto.go (hunk 3 above)
Comment (ssinha20) [reply pkg/crypto/crypto.go]: Please suggest.

Comment (harche) [reply pkg/crypto/crypto.go]: Throughout the execution of the `rakshctl` same key should be referred, and hence it's intentional the globally set value should be returned if it already exists. 

I can understand the inhibition to use global variables like that, maybe you can refactor that code to use closures but nevertheless, that function should always return the same key during the entire execution of `rakshctl`

Did you try deploying these changes on a system? 

@bpradipt @mkumatag

Comment (ssinha20) [reply pkg/crypto/crypto.go]: Please allow me to explain this again.
The function, "GetConfigMapKeys" gets called by both encrypt and decrypt functions.  Using global, as you explained above, is justified. 
However, the issue started happening when I passed wrong key value to decrypt function. The key passed is not null , but it's not the same one which was used to encrypt. Unless I removed this part, the return values were fixed, the global variables; not the ones which user passed.
Hence I wanted to get the user passed values being returned and commented this part. Waiting for comments from you all, @harche @bpradipt @mkumatag

File: pkg/crypto/crypto_test.go (hunk 4 above)
Comment (ssinha20) [reply pkg/crypto/crypto_test.go]: Please suggest if it would be fine to replace all panic calls with err calls in decrypt function? 
FYI, @harche @bpradipt @mkumatag

File: pkg/crypto/crypto.go (hunk 3 above)
Comment (ssinha20) [reply pkg/crypto/crypto.go]: I executed this as well. Without the commented out code, with wrong key too, decrypt worked.

File: pkg/crypto/crypto_test.go (hunk 1 above)
Comment (ssinha20) [reply pkg/crypto/crypto_test.go]: @harche: Please help with above query.

File: pkg/crypto/crypto.go (hunk 3 above)
Comment (harche) [reply pkg/crypto/crypto.go]: > Please allow me to explain this again.

I don't know what called for this. It's normal for reviewers to seek more information about the proposed changes.

Comment (harche) [reply pkg/crypto/crypto.go]: > The key passed is not null , but it's not the same one which was used to encrypt. Unless I removed this part, the return values were fixed, the global variables; not the ones which user passed.

This is an intended behavior. That function should return the same key (global) over and over again. Have you deployed your changes in the real cluster and tested the changes?

File: pkg/crypto/crypto_test.go (hunk 4 above)
Comment (harche) [reply pkg/crypto/crypto_test.go]: Let's deal with this once we have the clarity on why this change is needed.

File: pkg/crypto/crypto_test.go (hunk 1 above)
Comment (harche) [reply pkg/crypto/crypto_test.go:153]: I am not sure if this change is required at all.

Comment (harche) [reply pkg/crypto/crypto_test.go]: let's first understand if the modifying the `crypto.go` so that an unit test passes is a good idea. I would like to see the results of the actual deployment.

File: pkg/crypto/crypto_test.go (hunk 4 above)
Comment (harche) [reply pkg/crypto/crypto_test.go]: yes

File: pkg/crypto/crypto.go (hunk 3 above)
Comment (ssinha20) [reply pkg/crypto/crypto.go]: Thanks much for explaining the behavior of global key Harshal.  I will further explore the crypto functionalities and try to figure out correct condition.
Actions Items from my side
1.  correct the err code part for functions wherever applicable.
2.  revert changes for crypto.go for now.
3. fresh code drop to take care of above action items as part of this PR.

Please let me know if you have any other suggestions.

Comment (ssinha20) [reply pkg/crypto/crypto.go]: @harche : Can you please review latest code and let me know your comments.

**Review by harche — DISMISSED** [review]
LGTM

**Comment by harche** [comment]
@bpradipt @mkumatag not sure why CI is failing.

**Comment by mkumatag** [comment]
> @bpradipt @mkumatag not sure why CI is failing.

This is because of the go fmt issue:

```
$ make verify
./hack/verify-all.sh
./hack/..
Verify go-fmt!
diff -u ./pkg/crypto/crypto_test.go.orig ./pkg/crypto/crypto_test.go
--- ./pkg/crypto/crypto_test.go.orig	2020-06-04 11:08:17.396668930 +0000
+++ ./pkg/crypto/crypto_test.go	2020-06-04 11:08:17.396668930 +0000
@@ -108,4 +108,3 @@
 
 	return string(plaintextBytes)
 }
-
Run ./hack/update-gofmt.sh
Makefile:113: recipe for target 'verify' failed
make: *** [verify] Error 1
The command "make verify" exited with 2.
```

**Comment by mkumatag** [comment]
And if I click on the overall changes, I don't see any of the unit testcases anymore, not sure whats the problem!

**Comment by bpradipt** [comment]
I think it would be better to close this PR and re-submit a new PR.  I see the new test case `TestDecryptConfigMapWrongkey` deleted in the updated commit :-( or am I missing something ?

@harche @mkumatag ??

**Comment by harche** [comment]
@ssinha20 I was wondering for 3 lines of change which removes a redundant return of the error why do we need to have 8 commits? maybe we should squash those commits?

**Comment by harche** [comment]
> I think it would be better to close this PR and re-submit a new PR. I see the new test case `TestDecryptConfigMapWrongkey` deleted in the updated commit :-( or am I missing something ?

I don't see `TestDecryptConfigMapWrongkey` getting deleted. In the changed files I see only one return of error being removed correctly.

**Comment by ssinha20** [comment]
@harche :  All commits are squashed into single commit now. can you please review and let me know your comments for the same.

**Review by harche — APPROVED** [review]
LGTM

**Comment by ssinha20** [comment]
Regarding TestDecryptConfigMapWrongkey I have following work plan;
1.  handle in a separate PR. 
2.  add TestDecryptConfigMapWrongkey in crypto_test.go.  without changes in crypto.go; the TC   PASSED with wrong key.
Please suggest if this approach looks fine, then I will handle this in the new PR.

**Comment by bpradipt** [comment]
> Regarding TestDecryptConfigMapWrongkey I have following work plan;
> 
> 1. handle in a separate PR.
> 2. add TestDecryptConfigMapWrongkey in crypto_test.go.  without changes in crypto.go; the TC   PASSED with wrong key.
>    Please suggest if this approach looks fine, then I will handle this in the new PR.

Please open new PR for specific changes.

**Comment by ssinha20** [comment]
Sure Thanks.
